gpiozero/tones.py

```diff
@@ -0,0 +1,200 @@
+            A4_midi = 69
+            A4_freq = 440
+            return cls.from_frequency(A4_freq * 2 ** ((midi - A4_midi) / 12))
+        raise ValueError('invalid MIDI note: %r' % midi)
+
+    @classmethod
+    def from_note(cls, note):
+        """
+        Construct a :class:`Tone` from a musical note which must consist of
+        a capital letter A through G, followed by an optional semi-tone
+        modifier ("b" for flat, "#" for sharp, or their Unicode equivalents),
+        followed by an octave number (0 through 9).
+
+        For example `concert A`_, the typical tuning note at 440Hz, would be
+        represent as "A4". One semi-tone above this would be "A#4" or
+        alternatively "Bb4". Unicode representations of sharp and flat are also
+        accepted.
+        """
+        if isinstance(note, bytes):
+            note = note.decode('ascii')
+        if isinstance(note, str):
+            match = Tone.regex.match(note)
+            if match:
+                octave = int(match.group('octave')) + 1
+                return cls.from_midi(
+                    Tone.tones.index(match.group('note')) +
+                    Tone.semitones[match.group('semi')] +
+                    octave * 12)
+        raise ValueError('invalid note specification: %r' % note)
+
+    @classmethod
+    def from_frequency(cls, freq):
+        """
+        Construct a :class:`Tone` from a frequency specified in `Hz`_ which
+        must be a positive floating-point value in the range 0 < freq <= 20000.
+
+        .. _Hz: https://en.wikipedia.org/wiki/Hertz
+        """
+        if 0 < freq <= 20000:
+            return super(Tone, cls).__new__(cls, freq)
+        raise ValueError('invalid frequency: %.2f' % freq)
+
+    @property
+    def frequency(self):
+        """
+        Return the frequency of the tone in `Hz`_.
+
+        .. _Hz: https://en.wikipedia.org/wiki/Hertz
+        """
+        return float(self)
+
+    @property
+    def midi(self):
+        """
+        Return the (nearest) MIDI note to the tone's frequency. This will be an
+        integer number in the range 0 to 127. If the frequency is outside the
+        range represented by MIDI notes (which is approximately 8Hz to 12.5KHz)
+        :exc:`ValueError` exception will be raised.
+        """
+        result = int(round(12 * log2(self.frequency / 440) + 69))
+        if 0 <= result < 128:
+            return result
+        raise ValueError('%f is outside the MIDI note range' % self.frequency)
+
+    @property
+    def note(self):
+        """
+        Return the (nearest) note to the tone's frequency. This will be a
+        string in the form accepted by :meth:`from_note`. If the frequency is
+        outside the range represented by this format ("A0" is approximately
+        27.5Hz, and "G9" is approximately 12.5Khz) a :exc:`ValueError`
+        exception will be raised.
+        """
+        offset = self.midi - 60  # self.midi - A4_midi + Tone.tones.index('A')
```

Comment:
I wonder if it'd be better to do this calculation directly from `self.frequency`, instead of relying on `self.midi` ?

Reply:
It's possible but it'd just end up repeating the logarithm in the MIDI conversion. At the end of the day, both MIDI and the notation system are very similar: they both quantize the continuous frequency domain into a discrete "notes" domain. Looking at it another way: the MIDI system performs the quantization step, representing the result as a (limited) range of integers. The notation system then goes one step beyond this to a mod 12 representation (the octaves being the "major" step, and A..G, or rather C..B, being the "minor" steps).

That's without getting into the horror of different tuning systems though...

Reply:
Fair enough, I guess I was just worried about the different notation systems (possibly) having different min/max values.